nodejs/device/src/index.ts

```diff
@@ -0,0 +1,318 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+import { thiefSettings } from './thiefSettings'
+import { TimeoutError } from './timeout'
+import { Mqtt as ProvisioningTransport } from 'azure-iot-provisioning-device-mqtt';
+import { Mqtt as IotHubTransport } from 'azure-iot-device-mqtt';
+import { SymmetricKeySecurityClient } from 'azure-iot-security-symmetric-key';
+import { ProvisioningDeviceClient } from 'azure-iot-provisioning-device';
+import { Client, Message, Twin } from 'azure-iot-device';
+import { createHmac } from 'crypto';
+import { v4 as uuidv4 } from 'uuid';
+import { promisify } from 'util';
+
+const provisioningHost = process.env.THIEF_DEVICE_PROVISIONING_HOST;
+const idScope = process.env.THIEF_DEVICE_ID_SCOPE;
+const groupSymmetricKey = process.env.THIEF_DEVICE_GROUP_SYMMETRIC_KEY;
+const registrationId = process.env.THIEF_DEVICE_ID;
+const requestedServicePool = process.env.THIEF_REQUESTED_SERVICE_POOL;
+const runId = uuidv4();
+if (!provisioningHost || !idScope || !groupSymmetricKey || !registrationId || !requestedServicePool) {
+  throw new Error('Required environment variable is undefined.');
+}
+
+const settings: thiefSettings = {
+    thiefMaxRunDurationInSeconds: 0,
+    thiefPropertyUpdateIntervalInSeconds: 30,
+    thiefWatchdogFailureIntervalInSeconds: 300,
+    thiefAllowedClientLibraryExceptionCount: 10,
+    pairingRequestTimeoutIntervalInSeconds: 900,
+    pairingRequestSendIntervalInSeconds: 30,
+    sendMessageOperationsPerSecond: 1,
+    sendMessageAllowedFailureCount: 1000,
+    receiveC2dIntervalInSeconds: 20,
+    receiveC2dAllowedMissingMessageCount: 100,
+    reportedPropertiesUpdateIntervalInSeconds: 10,
+    reportedPropertiesAllowedFailureCount: 50,
+}
+
+class DeviceApp {
+  private async createDeviceClientUsingDpsGroupKey() {
+    const deviceKey = createHmac('SHA256', Buffer.from(groupSymmetricKey, 'base64')).update(registrationId, 'utf8').digest('base64');
+    const provisioningSecurityClient = new SymmetricKeySecurityClient(registrationId, deviceKey);
+    const provisioningClient = ProvisioningDeviceClient.create(provisioningHost, idScope, new ProvisioningTransport(), provisioningSecurityClient);
+    const registrationResultPromise = provisioningClient.register();
+    if (!registrationResultPromise) {
+      throw new Error('.register() did not return a promise when it should have.');
+    }
+    const registrationResult = await registrationResultPromise.catch(e => {throw new Error('DPS registration failed: ' + e.message)});
+    const connectionString = 'HostName=' + registrationResult.assignedHub + ';DeviceId=' + registrationResult.deviceId + ';SharedAccessKey=' + deviceKey;
+    const hubClient = Client.fromConnectionString(connectionString, IotHubTransport);
+    await hubClient.open().catch(e => {throw new Error('Opening client failed: ' + e.message)});
+    return hubClient;
+  }
+
+  private pairWithService() {
```

Comment:
>pairWithService [](start = 10, length = 15)

style question: if a function returns a Promise, should it be marked with the `async` keyword?  Because I think it should.  Otherwise I can't look at the signature and know what to do with the return value.  Or maybe it needs to be marked with the right typing to indicate that it returns a Promise.  That might be better.  Even though they're sorta the same thing, a linter would handle them differently.

Reply:
I didn't put a type annotation on it because TypeScript was already able to infer the return type as `Promise<void>`, but I'll add the return type to make it clear to the reader. 

Not sure if it's common to mark functions with the `async` keyword if it doesn't use `await`, I'll have to look into that.

Reply:
@YoDaMa @anthonyvercolano I asked in Teams, but might as well ask in here as well. Is it a good idea to mark the function with the `async` keyword if there is no use of `await`? Or should adding a type annotation for the return value be sufficient?